add basic regex test to detect invalid CompositeExplicitAutograd ops that are implemented with DispatchStub

bdhirsh opened the pull request:
This is a quick-and-dirty test that tries to parse the `aten/` folder to find ops that are implemented using DispatchStub, but are incorrectly marked as `CompositeExplicitAutograd` in native_functions.yaml. It should prevent bugs like https://github.com/pytorch/pytorch/issues/60892.

If we want to land this, I need to wait for <> to land to fix the existing bugs with `linalg_eigh` and `lu_solve`. The test correctly catches them and prints out:
```
AssertionError: 0 != 2 : Attempted to compare the lengths of [iterable] types: Expected: 0; Actual: 2.

Detected an operator that is marked as CompositeExplicitAutograd in native_functions.yaml, but appears to
be implemented using DispatchStub.

detected operators:
lu_solve
linalg_eigh

If you think that this is a false positive add the operator name to IGNORED_OPS in the file test/test_incorrect_dispatch_key_ops.py.
A common false positive is if you wrote your operator `at::native::{op}` as a wrapper that calls into
`at::native::{_op}` to do the real, backend-specific work. In this case, marking {op} with
CompositeExplicitAutograd and {_op} with CPU,CUDA is completely valid.
```

Quick note about the false positives: I added some more detail in the test comments, but an example false positive is `cumprod`. `at::native::cumprod` is correctly marked as `CompositeExplicitAutograd`, because (a) is has a derivative formula defined, and (b) `at::native::cumprod` is implemented by calling into `at::_cumprod` for the "real" work, making it backend-agnostic.

`_cumprod` is correctly marked as CPU/CUDA in native_functions.yaml, since it uses DispatchStub - but the function it calls is `cumprod_stub` - and since I'm relying on naming conventions to parse for problems, that shows up as a false positive.

Right now I have an allow list of false positives for the test to ignore.

Maybe... I should just update those false positive ops to follow the naming convention, e.g. call `_cumprod_stub`? Or detect the weird case where `{op}` is composite, but `_{op}` is not, and ignore that?

Either way, if this test goes into CI, there's some likelihood that it'll pick up false positives - The hope is that this catches enough bugs to be worth any of the flakiness.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61134 add basic regex test to detect invalid CompositeExplicitAutograd ops that are implemented with DispatchStub**
* #61133 fix eig.e bug - it uses DispatchStub and therefore isn't backend agnostic

Differential Revision: [D29526695](https://our.internmc.facebook.com/intern/diff/D29526695)

bdhirsh committed:
add basic regex test to detect invalid CompositeExplicitAutograd ops that are implemented with DispatchStub

[ghstack-poisoned]

bdhirsh committed:
Update on "add basic regex test to detect invalid CompositeExplicitAutograd ops that are implemented with DispatchStub"

This is a quick-and-dirty test that tries to parse the `aten/` folder to find ops that are implemented using DispatchStub, but are incorrectly marked as `CompositeExplicitAutograd` in native_functions.yaml. It should prevent bugs like https://github.com/pytorch/pytorch/issues/60892.

If we want to land this, I need to wait for <> to land to fix the existing bugs with `linalg_eigh` and `lu_solve`. The test correctly catches them and prints out:
```
AssertionError: 0 != 2 : Attempted to compare the lengths of [iterable] types: Expected: 0; Actual: 2.

Detected an operator that is marked as CompositeExplicitAutograd in native_functions.yaml, but appears to
be implemented using DispatchStub.

detected operators:
lu_solve
linalg_eigh

If you think that this is a false positive add the operator name to IGNORED_OPS in the file test/test_incorrect_dispatch_key_ops.py.
A common false positive is if you wrote your operator `at::native::{op}` as a wrapper that calls into
`at::native::{_op}` to do the real, backend-specific work. In this case, marking {op} with
CompositeExplicitAutograd and {_op} with CPU,CUDA is completely valid.
```

Quick note about the false positives: I added some more detail in the test comments, but an example false positive is `cumprod`. `at::native::cumprod` is correctly marked as `CompositeExplicitAutograd`, because (a) is has a derivative formula defined, and (b) `at::native::cumprod` is implemented by calling into `at::_cumprod` for the "real" work, making it backend-agnostic.

`_cumprod` is correctly marked as CPU/CUDA in native_functions.yaml, since it uses DispatchStub - but the function it calls is `cumprod_stub` - and since I'm relying on naming conventions to parse for problems, that shows up as a false positive.

Right now I have an allow list of false positives for the test to ignore.

Maybe... I should just update those false positive ops to follow the naming convention, e.g. call `_cumprod_stub`? Or detect the weird case where `{op}` is composite, but `_{op}` is not, and ignore that?

Either way, if this test goes into CI, there's some likelihood that it'll pick up false positives - The hope is that this catches enough bugs to be worth any of the flakiness.




[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29526695).

bdhirsh committed:
Update on "add basic regex test to detect invalid CompositeExplicitAutograd ops that are implemented with DispatchStub"

This is a quick-and-dirty test that tries to parse the `aten/` folder to find ops that are implemented using DispatchStub, but are incorrectly marked as `CompositeExplicitAutograd` in native_functions.yaml. It should prevent bugs like https://github.com/pytorch/pytorch/issues/60892.

If we want to land this, I need to wait for <> to land to fix the existing bugs with `linalg_eigh` and `lu_solve`. The test correctly catches them and prints out:
```
AssertionError: 0 != 2 : Attempted to compare the lengths of [iterable] types: Expected: 0; Actual: 2.

Detected an operator that is marked as CompositeExplicitAutograd in native_functions.yaml, but appears to
be implemented using DispatchStub.

detected operators:
lu_solve
linalg_eigh

If you think that this is a false positive add the operator name to IGNORED_OPS in the file test/test_incorrect_dispatch_key_ops.py.
A common false positive is if you wrote your operator `at::native::{op}` as a wrapper that calls into
`at::native::{_op}` to do the real, backend-specific work. In this case, marking {op} with
CompositeExplicitAutograd and {_op} with CPU,CUDA is completely valid.
```

Quick note about the false positives: I added some more detail in the test comments, but an example false positive is `cumprod`. `at::native::cumprod` is correctly marked as `CompositeExplicitAutograd`, because (a) is has a derivative formula defined, and (b) `at::native::cumprod` is implemented by calling into `at::_cumprod` for the "real" work, making it backend-agnostic.

`_cumprod` is correctly marked as CPU/CUDA in native_functions.yaml, since it uses DispatchStub - but the function it calls is `cumprod_stub` - and since I'm relying on naming conventions to parse for problems, that shows up as a false positive.

Right now I have an allow list of false positives for the test to ignore.

Maybe... I should just update those false positive ops to follow the naming convention, e.g. call `_cumprod_stub`? Or detect the weird case where `{op}` is composite, but `_{op}` is not, and ignore that?

Either way, if this test goes into CI, there's some likelihood that it'll pick up false positives - The hope is that this catches enough bugs to be worth any of the flakiness.


Differential Revision: [D29526695](https://our.internmc.facebook.com/intern/diff/D29526695)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29526695).

ezyang commented on test/test_incorrect_dispatch_key_ops.py:
```diff
@@ -0,0 +1,112 @@
+        curr_path = os.path.dirname(os.path.realpath(__file__))
+        root_dir = os.path.join(curr_path, '../aten/src/ATen/')
+        regex = re.compile("DEFINE_DISPATCH\((.*)_stub\)")
+        f_path = os.path.join(curr_path, '../aten/src/ATen/native/BinaryOps.cpp')
```
A little skeptical about hardcoding only BinaryOps.cpp file

ezyang replied:
Oh this code just looks dead

bdhirsh replied:
yep, dead :) whoops

ezyang commented on test/test_incorrect_dispatch_key_ops.py:
```diff
@@ -0,0 +1,112 @@
+        f_path = os.path.join(curr_path, '../aten/src/ATen/native/BinaryOps.cpp')
+
+        matches = []
+        for root, dirs, files in os.walk(root_dir):
```
Good hygiene for os.walk is to ignore private directories (directories staring with period). In this case it doesn't matter because you start traversing at aten/src/ATen, but failure to filter those directories out is a common cause of os.walk calls taking a really looooong time to run. Actually, even better would probably be to use `git ls-files` but that would require PyTorch to be a Git checkout (which is not necessarily true when people are running tests)

ezyang commented on test/test_incorrect_dispatch_key_ops.py:
```diff
@@ -0,0 +1,112 @@
+import os
```
It doesn't seem like this test is actually going to be run in CI

bdhirsh replied:
I was gonna be hopeful that we have glob magic somewhere for `test/*` and check the CI output. Thanks - I'll look for the right test file to add this file name.

ezyang commented on test/test_incorrect_dispatch_key_ops.py:
```diff
@@ -0,0 +1,112 @@
+                if invalid(op_name_str):
+                    invalid_op_entries.append(op_name_str)
+
+            elif isinstance(g, NativeFunctionsGroup):
```
A little confused here, don't you just want to run the check on a flat list of NativeFunctions (i.e. just don't bother grouping them)

bdhirsh replied:
I actually did that originally, but it causes the test to pick up way more false-positives.

For example, the native_functions.yaml entries for `at::abs` are:
```
- func: abs(Tensor self) -> Tensor
  dispatch:
    CompositeExplicitAutograd: abs
- func: abs_(Tensor(a!) self) -> Tensor(a!)
  dispatch:
    CompositeExplicitAutograd: abs_
- func: abs.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
  dispatch:
    CPU, CUDA: abs_out
```

It's valid for `abs` and `abs_` to be CompositeExplicitAutograd, and for `abs_out` to have backend-specific kernels that are implemented using DispatchStub. But the test would pick that up because it sees `DEFINE_DISPATCH(abs_stub`, along with a CompositeExplicitAutograd `abs` entry somewhere. This seemed like a common enough case that I just skip NativeFunctionGroup entries with a mix of composite and non-composite dispatch tables.

bdhirsh replied:
Hey, maybe we could enforce a naming convention (through this test) that if an `{op}_out` kernel uses DispatchStub, it has to name the corresponding dispatch kernel `{op}_out_stub`. But that just seems even more intrusive.

ezyang reviewed (approved):
Sure, let's give this a try. Please see comments

bdhirsh commented:
I mucked with git incorrectly locally and somehow can't find the commit corresponding to this PR on my local branch - ghstack made a new PR here instead: https://github.com/pytorch/pytorch/pull/61295. 😬 

I addressed all of the feedback there (minus the comment about how I dealt with `NativeFunctionsGroup`, see the comment I left).
